internal/utils/validation/services.go

```diff
@@ -29,17 +32,112 @@ import (
+		errs = errors.Join(errs, validateServiceTemplate(ctx, cl, svc, ns))
+		if svc.TemplateChain == "" {
 			continue
 		}
+		errs = errors.Join(errs, validateServiceTemplateChain(ctx, cl, svc, ns))
+	}
+
+	return errs
+}
 
-		if !svcTemplate.Status.Valid {
-			errs = errors.Join(errs, fmt.Errorf("the ServiceTemplate %s is invalid with the error: %s", key, svcTemplate.Status.ValidationError))
+// validateServiceTemplate validates the given [github.com/K0rdent/kcm/api/v1alpha1.ServiceTemplate] checking if it is valid
+func validateServiceTemplate(ctx context.Context, cl client.Client, svc kcmv1.Service, ns string) error {
+	svcTemplate := new(kcmv1.ServiceTemplate)
+	key := client.ObjectKey{Namespace: ns, Name: svc.Template}
+	if err := cl.Get(ctx, key, svcTemplate); err != nil {
+		return fmt.Errorf("failed to get ServiceTemplate %s: %w", key, err)
+	}
+
+	if !svcTemplate.Status.Valid {
+		return fmt.Errorf("the ServiceTemplate %s is invalid with the error: %s", key, svcTemplate.Status.ValidationError)
+	}
+
+	return nil
+}
+
+// validateServiceTemplateChain validates the given [github.com/K0rdent/kcm/api/v1alpha1.ServiceTemplateChain] checking if
+// it contains valid [github.com/K0rdent/kcm/api/v1alpha1.ServiceTemplate] with matching version.
+func validateServiceTemplateChain(ctx context.Context, cl client.Client, svc kcmv1.Service, ns string) error {
+	templateChain := new(kcmv1.ServiceTemplateChain)
+	key := client.ObjectKey{Namespace: ns, Name: svc.TemplateChain}
+	if err := cl.Get(ctx, key, templateChain); err != nil {
+		return fmt.Errorf("failed to get ServiceTemplateChain %s: %w", key, err)
+	}
+
+	if !templateChain.Status.Valid {
+		return fmt.Errorf("the ServiceTemplateChain %s is invalid with the error: %s", key, templateChain.Status.ValidationError)
+	}
+
+	var errs error
+	matchingTemplateFound := false
+	for _, t := range templateChain.Spec.SupportedTemplates {
+		if t.Name != svc.Template {
+			continue
 		}
+		template := new(kcmv1.ServiceTemplate)
+		key = client.ObjectKey{Namespace: ns, Name: t.Name}
+		if err := cl.Get(ctx, key, template); err != nil {
+			errs = errors.Join(errs, fmt.Errorf("failed to get ServiceTemplate %s: %w", key, err))
+			continue
+		}
+		// this error should never happen, but we check it anyway
+		if !template.Status.Valid {
+			errs = errors.Join(errs, fmt.Errorf("the ServiceTemplate %s is invalid with the error: %s", key, template.Status.ValidationError))
+			continue
+		}
+		matchingTemplateFound = true
+		break
+	}
+	if !matchingTemplateFound {
+		errs = errors.Join(errs, fmt.Errorf("the ServiceTemplateChain %s does not support ServiceTemplate %s", key, svc.Template))
 	}
 
 	return errs
 }
+
+func ValidateUpgradePaths(services []kcmv1.Service, upgradePaths []kcmv1.ServiceUpgradePaths) error {
+	observedTemplatesMap := make(map[string]struct {
+		Template     string
+		UpgradePaths [][]string
+	}, len(upgradePaths))
```

Comment:
We could also do the following instead of creating a new struct type:
```go
	observedTemplatesMap := make(map[string]kcmv1.ServiceUpgradePaths, len(upgradePaths))
```

Reply:
I meant we could do away with the anonymous struct and just re-use `kcmv1.ServiceUpgradePaths`, it's up to you.